### ISSUE TITLE
Spontaneously crashing notebooks: "zmq.error.ZMQError: Invalid argument" 

### ISSUE THREAD

**Issue by bighomeyg** [jupyter/notebook#1906]
My Jupyter notebooks are spontaneously crashing and returning the error posted below. When the crash happens, I lose the connection to the kernel entirely and all unsaved progress is lost. I can't seem to isolate the error to any particular behavior (same result with Chrome and Safari, different scripts loaded in the notebook, etc.). Sometimes it will crash within the first 30 seconds, other times it will take an hour before crashing, but it's crashing 100% of my notebooks. I'm running on OSX 10.11 with a macports installation. The issue started a few days ago, and I haven't found any changes pushed to macports for zmq or jupyter that are concurrent with the issue. Any help is much appreciated!

```
  File "/opt/local/bin/jupyter-notebook-2.7", line 11, in <module>
    load_entry_point('notebook==4.2.3', 'console_scripts', 'jupyter-notebook')()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/jupyter_core/application.py", line 267, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/traitlets/config/application.py", line 658, in launch_instance
    app.start()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/notebook/notebookapp.py", line 1162, in start
    self.io_loop.start()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/zmq/eventloop/ioloop.py", line 183, in start
    raise e
zmq.error.ZMQError: Invalid argument
```

**Comment by andreavicere** [comment]
I am having exactly the same problem, since about one week. The only difference is that the I am using jupyter on macOS 10.12.1.
It "might" be related to the latest update of zmq: it has been released on November 4th, and I may have updated (using macports) to this version roughly one week ago.
Some notebooks that were running fine started crashing, independently on the browser or the status of the machine.

**Comment by andreavicere** [comment]
An update: although I've not accumulated large statistics, it seems that jupyter notebook does not crash if started on a _specific_ file.
In other words, if I launch
`jupyter notebook myfile.ipynb`
all runs fine.
If instead I launch as
`jupyter notebook`
and then I pick the _same_ file from the directory, after a while the usual crash occurs.
I thought to report, as a possible way to keep working while the issue gets pinned down.

**Comment by takluyver** [comment]
Thanks for the info! I can't think how that would make any difference to ZMQ, but maybe it will suggest something to Min.

**Comment by bighomeyg** [comment]
Thanks for the tip! I'll test it out and let you know if that works for me. I've been using jupyter notebooks on a different machine (OSX 10.12, macports install), but with an older version of zmq (4.1.5) and I haven't had any crashes. Looks like this might be specific to zmq 4.2.

**Comment by andreavicere** [comment]
Thanks for confirming that jupyter is stable with zmq 4.1.5 !
It is possible to revert to an older port, following the instructions here:
https://trac.macports.org/wiki/howto/InstallingOlderPort
The only difficulty for me was determining the revision number: for zmq 4.1.5 one needs
svn checkout -r 149404 https://svn.macports.org/repository/macports/trunk/dports/devel/zmq
Incidentally, I thought that a version 4.1.6 existed, but couldn't find the tag.

**Comment by minrk** [comment]
Thanks for the report! It sounds like this is a bug in the recently released libzmq-4.2.0. The relevant call is likely in `zmq_poll`, which has changed quite a bit in zmq 4.2.

**Comment by jwoillez** [comment]
FYI: The update by macports of py-zmq@16.0.2 fixes the issue for me. Thanks @minrk.

**Comment by minrk** [comment]
Thanks, @jwoillez!

**Comment by andreavicere** [comment]
I also confirm that the issue is fixed with py-zmq@16.0.2, as available now from macports. Awesome! Big thanks @minrk !

**Comment by bighomeyg** [comment]
Thanks so much for the quick fix, @minrk !
